source/module_elecstate/potentials/efield.cpp

```diff
@@ -303,4 +306,57 @@ void Efield::prepare(const UnitCell &cell, double &latvec, double &area)
     bmod = sqrt(pow(bvec[0], 2) + pow(bvec[1], 2) + pow(bvec[2], 2));
 }
 
+void Efield::autoset(const UnitCell& cell)
+{
+    if (efield_pos_max == -1 || efield_pos_dec == -1)
+    {
+        // obtain the position of atoms along the efield direction
+        std::vector<double> pos;
+        for (int it = 0; it < cell.ntype; ++it)
+        {
+            for (int ia = 0; ia < cell.atoms[it].na; ++ia)
+            {
+                pos.push_back(cell.atoms[it].taud[ia][efield_dir]);
```

Comment:
Up to now, the use of member(s) in `GlobalC` is quite sensitive, do you think it would be better if only feed in the `std::vector<double> pos`, rather than the reference of `UnitCell`? I think the direct use of `UnitCell` here also will brings trouble when you write unittest for it. But if you find it is quite inconvienent to get rid of `UnitCell` inside this function, and when you want to add unittest for it, there is a file mock_unitcell.cpp might be helpful.

Reply:
OK, I will modify it.